format.swift

```diff
@@ -0,0 +1,548 @@
+
+  init(
+    executableURL: URL,
+    arguments: [String],
+    additionalEnvironment: [String: String] = [:]
+  ) {
+    process = Process()
+    process.executableURL = executableURL
+    process.arguments = arguments
+    process.environment = additionalEnvironment.merging(ProcessInfo.processInfo.environment) { (additional, _) in additional }
+  }
+
+  @discardableResult
+  func run(
+    verbose: Bool,
+    captureStdout: Bool = true,
+    captureStderr: Bool = true
+  ) throws -> ProcessResult {
+    if verbose {
+      logProcessCommand(executableURL: process.executableURL, arguments: process.arguments)
+    }
+
+    let group = DispatchGroup()
+
+    var stdoutData = Data()
+    if captureStdout {
+      let outPipe = Pipe()
+      process.standardOutput = outPipe
+      addHandler(pipe: outPipe, group: group) { stdoutData.append($0) }
+    }
+
+    var stderrData = Data()
+    if captureStderr {
+      let errPipe = Pipe()
+      process.standardError = errPipe
+      addHandler(pipe: errPipe, group: group) { stderrData.append($0) }
+    }
+
+    try process.run()
+    SigIntListener.runningSubprocesses.insert(process)
+    process.waitUntilExit()
+    SigIntListener.runningSubprocesses.remove(process)
+
+    if captureStdout || captureStderr {
+      // Make sure we've received all stdout/stderr
+      group.wait()
+    }
+
+    guard let stdoutString = String(data: stdoutData, encoding: .utf8) else {
+      throw FailedToDecodeUTF8Error(data: stdoutData)
+    }
+    guard let stderrString = String(data: stderrData, encoding: .utf8) else {
+      throw FailedToDecodeUTF8Error(data: stderrData)
+    }
+
+    guard process.terminationStatus == 0 else {
+      throw NonZeroExitCodeError(
+        stdout: stdoutString,
+        stderr: stderrString,
+        exitCode: Int(process.terminationStatus)
+      )
+    }
+
+    return ProcessResult(
+      stdout: stdoutString,
+      stderr: stderrString
+    )
+  }
+
+  private func addHandler(
+    pipe: Pipe,
+    group: DispatchGroup,
+    addData: @escaping (Data) -> Void
+  ) {
+    group.enter()
+    pipe.fileHandleForReading.readabilityHandler = { fileHandle in
+      // Apparently using availableData can cause various issues
+      let newData = fileHandle.readData(ofLength: Int.max)
+      if newData.count == 0 {
+        pipe.fileHandleForReading.readabilityHandler = nil;
+        group.leave()
+      } else {
+        addData(newData)
+      }
+    }
+  }
+}
+
+/// The exit code and output (if redirected) from a subprocess that has
+/// terminated
+struct ProcessResult {
+  let stdout: String
+  let stderr: String
+}
+
+/// Error thrown if a process terminates with a non-zero exit code.
+struct NonZeroExitCodeError: Error {
+  let stdout: String
+  let stderr: String
+  let exitCode: Int
+}
+
+/// Error thrown if `stdout` or `stderr` could not be decoded as UTF-8.
+struct FailedToDecodeUTF8Error: Error {
+  let data: Data
+}
+
+// MARK: Actual implementation
```

Comment:
Could you flip this? ie. implementation at the start of the file, copied junk that we don't care about at the end?